utils/div_ceil.hh

```diff
@@ -9,8 +9,25 @@
 #pragma once
 
 #include <concepts>
+#include <limits>
+#include "utils/assert.hh"
 
+template <std::integral Divdend, std::integral Divisor>
 inline auto
-div_ceil(std::integral auto dividend, std::integral auto divisor) {
+div_ceil(Divdend dividend, Divisor divisor) {
+    SCYLLA_ASSERT(divisor != 0);
+    using common_t = std::common_type_t<Divdend, Divisor>;
+    common_t a = static_cast<common_t>(dividend);
+    common_t b = static_cast<common_t>(divisor);
+    if (b > 0) {
+        // check for overflow in (dividend + divisor)
+        SCYLLA_ASSERT(a <= std::numeric_limits<common_t>::max() - b);
+        // check for underflow in (dividend + divisor - 1)
+        SCYLLA_ASSERT(a + b > std::numeric_limits<common_t>::min());
+    } else {
+        // check for underflow in (dividend + divisor)
+        SCYLLA_ASSERT(a >= std::numeric_limits<common_t>::min() - b);
+        // no need to check for overflow in (dividend + divisor - 1), as divisor < 0
+    }
     return (dividend + divisor - 1) / divisor;
```

Comment:
- With `dividend=INT_MIN+1` and `divisor=(-1)`, we still get undefined behavior. These inputs are both `int`, the common type is `int`, therefore `b` is an `int` with value `-1`. The second branch takes effect, and we pass the assert (because `INT_MIN+1 >= INT_MIN - (-1)`). However, on the final line, we calculate the sum `INT_MIN+1 + (-1) - 1`, which is `INT_MIN - 1`; that's UB.
- With `dividend=INT_MIN+2` and `divisor=(-1)`, we also get undefined behavior, namely from the division. On the last line, we get `(INT_MIN+2 + (-1) - 1) / (-1)`, which is `INT_MIN / (-1)`; and on a platform using two's complement representation, the result is a positive value that cannot be represented as an `int` (namely `INT_MAX+1`).


As I mention above, IMO this function should only ever be called with a nonnegative dividend and a positive divisor. I expect we want to divide
and round up only such quantities where negative values make no sense; sizes, counts, durations, and so on. Therefore I propose:

~~~
    SCYLLA_ASSERT(dividend >= 0);
    SCYLLA_ASSERT(divisor > 0);
    Divisor max_remainder = divisor - 1;
    using lim = std::numeric_limits<std::common_type_t<Divdend, Divisor>>
    SCYLLA_ASSERT(dividend <= lim::max() - max_remainder);
    return (dividend + max_remainder) / divisor;
~~~